# 1. Import

In [1]:
import pandas as pd
import numpy as np
import nycflights13 as flights


# 2. Data load

In [2]:
# 항공편 데이터 (main dataset)
df_flights = flights.flights
df_airlines = flights.airlines
df_airports = flights.airports
df_planes = flights.planes
df_weather = flights.weather

# 3. Data Analysis

## 3-1. flights



In [3]:
df_flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            336776 non-null  int64  
 1   month           336776 non-null  int64  
 2   day             336776 non-null  int64  
 3   dep_time        328521 non-null  float64
 4   sched_dep_time  336776 non-null  int64  
 5   dep_delay       328521 non-null  float64
 6   arr_time        328063 non-null  float64
 7   sched_arr_time  336776 non-null  int64  
 8   arr_delay       327346 non-null  float64
 9   carrier         336776 non-null  object 
 10  flight          336776 non-null  int64  
 11  tailnum         334264 non-null  object 
 12  origin          336776 non-null  object 
 13  dest            336776 non-null  object 
 14  air_time        327346 non-null  float64
 15  distance        336776 non-null  int64  
 16  hour            336776 non-null  int64  
 17  minute    

In [4]:
df_flights.columns

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour'],
      dtype='object')

## 3-2. airlines

In [5]:
df_airlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   carrier  16 non-null     object
 1   name     16 non-null     object
dtypes: object(2)
memory usage: 388.0+ bytes


## 3-3. airports

In [6]:
df_airports.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458 entries, 0 to 1457
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   faa     1458 non-null   object 
 1   name    1458 non-null   object 
 2   lat     1458 non-null   float64
 3   lon     1458 non-null   float64
 4   alt     1458 non-null   int64  
 5   tz      1458 non-null   int64  
 6   dst     1458 non-null   object 
 7   tzone   1455 non-null   object 
dtypes: float64(2), int64(2), object(4)
memory usage: 91.3+ KB


## 3-4. planes

In [7]:
df_planes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3322 entries, 0 to 3321
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   tailnum       3322 non-null   object 
 1   year          3252 non-null   float64
 2   type          3322 non-null   object 
 3   manufacturer  3322 non-null   object 
 4   model         3322 non-null   object 
 5   engines       3322 non-null   int64  
 6   seats         3322 non-null   int64  
 7   speed         23 non-null     float64
 8   engine        3322 non-null   object 
dtypes: float64(2), int64(2), object(5)
memory usage: 233.7+ KB


## 3-5. weather

### 데이터 기본 정보 확인

In [8]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26115 entries, 0 to 26114
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   origin      26115 non-null  object 
 1   year        26115 non-null  int64  
 2   month       26115 non-null  int64  
 3   day         26115 non-null  int64  
 4   hour        26115 non-null  int64  
 5   temp        26114 non-null  float64
 6   dewp        26114 non-null  float64
 7   humid       26114 non-null  float64
 8   wind_dir    25655 non-null  float64
 9   wind_speed  26111 non-null  float64
 10  wind_gust   5337 non-null   float64
 11  precip      26115 non-null  float64
 12  pressure    23386 non-null  float64
 13  visib       26115 non-null  float64
 14  time_hour   26115 non-null  object 
dtypes: float64(9), int64(4), object(2)
memory usage: 3.0+ MB


In [9]:
df_weather.shape

(26115, 15)

In [10]:
df_weather.head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  2013      1    1     5  39.02  28.04  64.43     260.0    12.65858   

   wind_gust  precip  pressure  visib             time_hour  
0        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z

In [11]:
df_weather.columns

Index(['origin', 'year', 'month', 'day', 'hour', 'temp', 'dewp', 'humid',
       'wind_dir', 'wind_speed', 'wind_gust', 'precip', 'pressure', 'visib',
       'time_hour'],
      dtype='object')

'origin' - 기상 관측소 코드

'year' 'month', 'day', 'hour' - 기록 시점

'temp' 기온 (F) - 물이 어는 온도는 32도(섭씨 0도) / 물이 끓는 온도는 212도(섭씨 100도)

'dewp' 이슬점 (F) - 공기 중의 수증기가 포화 상태에 도달하여 이슬이나 안개 형태로 응결되기 시작하는 온도

'humid' 상대습도 - 기온에 따른 습하고 건조한 정도를 백분율로 나타낸 것

'wind_dir' 풍향 - 동풍 90도 / 남풍 180도 / 서풍 270도 / 북향 360도

'wind_speed' 풍속 (mph(mile per hour))

'wind_gust' 돌풍 속도 (mph) - (결측치 많음 20778/26115)

'precip' 강수량 inch (1 inch == 25.4 mm)

'pressure' 해수면 기압 (mb)

'visib' 가시거리 (mile)

'time_hour' POSIXct 형식의 기록 시각

### 결측치 확인

In [12]:
df_weather.isna().sum()

origin            0
year              0
month             0
day               0
hour              0
temp              1
dewp              1
humid             1
wind_dir        460
wind_speed        4
wind_gust     20778
precip            0
pressure       2729
visib             0
time_hour         0
dtype: int64

In [13]:
df_weather[~df_weather["wind_gust"].isna()].head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
14    EWR  2013      1    1    16  37.04  19.94  49.62     300.0    13.80936   
16    EWR  2013      1    1    18  33.98  15.08  45.43     310.0    12.65858   
19    EWR  2013      1    1    21  30.02  12.92  48.48     320.0    18.41248   
20    EWR  2013      1    1    22  28.94  12.02  48.69     320.0    18.41248   
22    EWR  2013      1    2     0  26.96  10.94  50.34     310.0    14.96014   

    wind_gust  precip  pressure  visib             time_hour  
14   20.71404     0.0    1012.1   10.0  2013-01-01T21:00:00Z  
16   25.31716     0.0    1014.1   10.0  2013-01-01T23:00:00Z  
19   26.46794     0.0    1016.0   10.0  2013-01-02T02:00:00Z  
20   25.31716     0.0    1016.5   10.0  2013-01-02T03:00:00Z  
22   25.31716     0.0    1016.3   10.0  2013-01-02T05:00:00Z

In [14]:
df_weather[df_weather["wind_gust"].isna()].tail()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
26104    LGA  2013     12   30     8  44.06  33.98  67.45     260.0   
26105    LGA  2013     12   30     9  44.06  33.08  65.07     260.0   
26106    LGA  2013     12   30    10  42.98  33.80  70.28     330.0   
26113    LGA  2013     12   30    17  30.92  12.92  46.74     320.0   
26114    LGA  2013     12   30    18  28.94  10.94  46.41     330.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
26104    11.50780        NaN     0.0    1012.9   10.0  2013-12-30T13:00:00Z  
26105    13.80936        NaN     0.0    1013.7   10.0  2013-12-30T14:00:00Z  
26106    16.11092        NaN     0.0       NaN   10.0  2013-12-30T15:00:00Z  
26113    17.26170        NaN     0.0    1019.9   10.0  2013-12-30T22:00:00Z  
26114    18.41248        NaN     0.0    1020.9   10.0  2013-12-30T23:00:00Z

### 고유값 확인

In [15]:
df_weather.shape

(26115, 15)

In [16]:
df_weather.nunique()

origin           3
year             1
month           12
day             31
hour            24
temp           173
dewp           153
humid         2499
wind_dir        37
wind_speed      36
wind_gust       37
precip          59
pressure       468
visib           20
time_hour     8714
dtype: int64

### 통계값 확인

In [17]:
df_weather.groupby("origin").mean(numeric_only=True)

year     month        day       hour       temp       dewp  \
origin                                                                 
EWR     2013.0  6.503505  15.672642  11.488337  55.546553  41.852156   
JFK     2013.0  6.503676  15.678612  11.490811  54.472150  41.857119   
LGA     2013.0  6.504020  15.674707  11.493223  55.762605  40.610868   

            humid    wind_dir  wind_speed  wind_gust    precip     pressure  \
origin                                                                        
EWR     63.062162  195.483604    9.461073  24.135727  0.005042  1017.832801   
JFK     65.205077  204.183709   11.468396  27.563739  0.003985  1018.180698   
LGA     59.323183  199.510113   10.622839  25.144656  0.004381  1017.678161   

           visib  
origin            
EWR     9.277015  
JFK     9.177578  
LGA     9.311531

In [18]:
grouped_df = df_weather.groupby(["origin","year","month"]).mean(numeric_only=True)
grouped_df

day       hour       temp       dewp      humid  \
origin year month                                                          
EWR    2013 1      16.040431  11.514825  35.562156  22.672507  62.124515   
            2      14.478326  11.496263  34.263318  21.936143  63.335575   
            3      16.008075  11.512786  40.118654  25.032571  57.913822   
            4      15.500000  11.500000  52.977500  34.574250  54.235125   
            5      16.000000  11.500000  63.320242  49.237419  64.073710   
            6      15.500000  11.500000  73.267250  60.398750  67.002292   
            7      16.017544  11.522267  80.702996  67.525830  66.676410   
            8      15.971622  11.459459  74.537483  61.156590  65.751746   
            9      15.518776  11.493741  67.304784  53.592990  64.035563   
            10     15.904891  11.461957  59.778207  47.017038  65.319280   
            11     15.591608  11.447552  44.577343  28.788448  56.520112   
            12     15.396359  11.449580  37.950084  28.198571  69.721415   
JFK    2013 1      16.040431  11.514825  35.385553  22.332884  61.661617   
            2      14.490313  11.517139  34.192459  21.570462  62.558480   
            3      16.022911  11.526954  39.544717  24.620970  58.022224   
            4      15.518776  11.488178  50.142698  34.705508  60.057636   
            5      16.000000  11.500000  59.314758  48.650726  70.989261   
            6      15.500000  11.500000  69.958250  60.286750  73.761472   
            7      16.000000  11.500000  78.734919  67.903468  71.744691   
            8      15.975610  11.490515  73.818780  61.526098  67.844201   
            9      15.500000  11.500000  66.897750  53.673000  65.251819   
            10     15.924119  11.448509  59.801951  47.023171  65.063618   
            11     15.632539  11.469846  45.134194  29.447938  56.799607   
            12     15.398601  11.433566  38.604867  28.347133  68.313413   
LGA    2013 1      16.033693  11.522911  35.959272  22.037655  59.156092   
            2      14.479104  11.502985  34.356119  21.155403  60.686030   
            3      16.022911  11.526954  39.976523  24.694474  56.559474   
            4      15.500000  11.500000  52.114500  33.597750  53.150597   
            5      16.000000  11.500000  62.750000  47.610403  61.496183   
            6      15.500000  11.500000  73.326500  57.939750  61.226861   
            7      15.979812  11.512786  80.764253  65.587806  62.264038   
            8      15.983762  11.499323  75.048254  59.397267  60.345737   
            9      15.500000  11.500000  67.911250  52.463000  59.677431   
            10     15.924119  11.448509  60.632439  45.826585  60.138753   
            11     15.632539  11.469846  45.260926  28.233128  53.492959   
            12     15.398601  11.433566  38.769762  26.615105  63.627692   

                     wind_dir  wind_speed  wind_gust    precip     pressure  \
origin year month                                                             
EWR    2013 1      217.675378    9.874685  27.321978  0.004757  1020.977557   
            2      206.396947   12.202740  26.227938  0.005725  1016.240102   
            3      236.331034   11.613262  24.827094  0.004038  1013.533481   
            4      170.686695    9.628193  23.468091  0.002042  1020.506240   
            5      180.098039    8.488549  23.742829  0.007312  1018.396880   
            6      193.304598    9.551474  22.339851  0.012125  1013.570947   
            7      194.439776    9.148779  21.168017  0.005047  1016.555728   
            8      196.557610    7.623140  21.171579  0.006176  1016.641496   
            9      191.979914    8.029852  21.764266  0.002142  1016.996280   
            10     150.378151    8.324392  22.434833  0.000679  1018.108072   
            11     209.455587   10.348973  26.641176  0.004168  1022.539024   
            12     198.070922    8.932245  23.570193  0.006373  1019.880363   
JFK    2013 1      235.829960 